llvm/include/llvm/CodeGen/SDPatternMatch.h

```diff
@@ -0,0 +1,712 @@
+    return true;
+  }
+};
+
+template <typename Pred, typename... Preds>
+struct And<Pred, Preds...> : And<Preds...> {
+  Pred P;
+  And(Pred &&p, Preds &&...preds)
+      : And<Preds...>(std::forward<Preds>(preds)...), P(std::forward<Pred>(p)) {
+  }
+
+  template <typename MatchContext>
+  bool match(const MatchContext &Ctx, SDValue N) {
+    return P.match(Ctx, N) && And<Preds...>::match(Ctx, N);
+  }
+};
+
+template <typename... Preds> struct Or {
+  template <typename MatchContext> bool match(const MatchContext &, SDValue N) {
+    return false;
+  }
+};
+
+template <typename Pred, typename... Preds>
+struct Or<Pred, Preds...> : Or<Preds...> {
+  Pred P;
+  Or(Pred &&p, Preds &&...preds)
+      : Or<Preds...>(std::forward<Preds>(preds)...), P(std::forward<Pred>(p)) {}
+
+  template <typename MatchContext>
+  bool match(const MatchContext &Ctx, SDValue N) {
+    return P.match(Ctx, N) || Or<Preds...>::match(Ctx, N);
+  }
+};
+
+template <typename... Preds> And<Preds...> m_all_of(Preds &&...preds) {
+  return And<Preds...>(std::forward<Preds>(preds)...);
+}
+
+template <typename... Preds> Or<Preds...> m_any_of(Preds &&...preds) {
+  return Or<Preds...>(std::forward<Preds>(preds)...);
+}
+
+// === Generic node matching ===
+template <typename... OpndPreds> struct Node_match {
+  unsigned Opcode;
+  unsigned OpIdx;
+
+  Node_match(unsigned Opc, unsigned OpIdx) : Opcode(Opc), OpIdx(OpIdx) {}
+
+  template <typename MatchContext>
+  bool match(const MatchContext &Ctx, SDValue N) {
+    if (!N)
+      return false;
+
+    if (OpIdx == 0) {
+      // Check opcode
+      if (!sd_context_match(N, Ctx, m_Opc(Opcode)))
+        return false;
+    }
+
+    // Returns false if there are more operands than predicates;
+    return N->getNumOperands() == OpIdx;
+  }
+};
+
+template <typename OpndPred, typename... OpndPreds>
+struct Node_match<OpndPred, OpndPreds...> : Node_match<OpndPreds...> {
+  unsigned Opcode;
+  unsigned OpIdx;
+  OpndPred P;
+
+  Node_match(unsigned Opc, unsigned OpIdx, OpndPred &&p, OpndPreds &&...preds)
+      : Node_match<OpndPreds...>(Opc, OpIdx + 1,
+                                 std::forward<OpndPreds>(preds)...),
+        Opcode(Opc), OpIdx(OpIdx), P(std::forward<OpndPred>(p)) {}
+
+  template <typename MatchContext>
+  bool match(const MatchContext &Ctx, SDValue N) {
+    if (!N)
+      return false;
+
+    if (OpIdx == 0) {
+      // Check opcode
+      if (!sd_context_match(N, Ctx, m_Opc(Opcode)))
+        return false;
+    }
+
+    if (OpIdx < N->getNumOperands())
+      return P.match(Ctx, N->getOperand(OpIdx)) &&
+             Node_match<OpndPreds...>::match(Ctx, N);
+
+    // This is the case where there are more predicates than operands.
+    return false;
+  }
+};
+
+template <typename... OpndPreds>
+Node_match<OpndPreds...> m_Node(unsigned Opcode, OpndPreds &&...preds) {
+  return Node_match<OpndPreds...>(Opcode, 0, std::forward<OpndPreds>(preds)...);
+}
```

Comment:
I didn't even try to understand this.

Reply:
This is trying to generically match a node by its opcode and operands using variadic template argument tricks. The implementation is similar to `m_AllOf` and `m_AnyOf` above.

Reply:
I find the implementation of this one quite hard to follow, compared to m_AllOg/m_AnyOf. Especially the presence of Opcode and OpIndex as members in each recursively inherited struct seems odd. OpIndex should probably be a template parameter, and Opcode checked separately from the rest.

Reply:
> I find the implementation of this one quite hard to follow, compared to m_AllOg/m_AnyOf. Especially the presence of Opcode and OpIndex as members in each recursively inherited struct seems odd. OpIndex should probably be a template parameter, and Opcode checked separately from the rest.

I've simplified this by only doing recursive template matching on operands and check the opcode in `m_Node`.